optuna/samplers/_tpe/parzen_estimator.py

```diff
@@ -18,114 +33,406 @@ class _ParzenEstimatorParameters(
+
+        pairs_of_observation_and_idx = np.asarray(
+            [(low, -1)] + [(x, i) for i, x in enumerate(observations)] + [(high, n_observations)]
+        )
+        pairs_of_observation_and_idx = pairs_of_observation_and_idx[
+            np.argsort(pairs_of_observation_and_idx[:, 0])
+        ]
 
-        # We decide the weights.
-        unsorted_weights = weights_func(mus.size)
         if consider_prior:
-            sorted_weights = numpy.zeros_like(sorted_mus)
-            sorted_weights[:prior_pos] = unsorted_weights[order[:prior_pos]]
-            sorted_weights[prior_pos] = prior_weight
-            sorted_weights[prior_pos + 1 :] = unsorted_weights[order[prior_pos:]]
+
+            prior_mu = 0.5 * (low + high)
+            prior_sigma = 1.0 * (high - low)
+
+            mus = np.empty(n_observations + 1)
+            mus[:n_observations] = observations
+            mus[n_observations] = prior_mu
+
+            if sigmas0 is None:
+                pairs_of_observation_and_idx = np.vstack(
+                    [pairs_of_observation_and_idx, (prior_mu, n_observations + 1)]
+                )
+                pairs_of_observation_and_idx = pairs_of_observation_and_idx[
+                    np.argsort(pairs_of_observation_and_idx[:, 0])
+                ]
+
+                pairs_of_sigma_and_idx = np.empty((n_observations + 1, 2))
+                pairs_of_sigma_and_idx[:, 0] = np.maximum(
+                    pairs_of_observation_and_idx[1:-1, 0] - pairs_of_observation_and_idx[0:-2, 0],
+                    pairs_of_observation_and_idx[2:, 0] - pairs_of_observation_and_idx[1:-1, 0],
+                )
+                pairs_of_sigma_and_idx[:, 1] = pairs_of_observation_and_idx[1:-1, 1]
+
+                sigmas = np.empty(n_observations + 1)
+                sigmas[:n_observations] = pairs_of_sigma_and_idx[
+                    np.argsort(pairs_of_sigma_and_idx[:, 1])
+                ][:-1, 0]
+                sigmas[n_observations] = prior_sigma
+            else:
+                sigmas = np.empty(n_observations + 1)
+                sigmas[:n_observations] = sigmas0 * (high - low)
+                sigmas[n_observations] = prior_sigma
         else:
-            sorted_weights = unsorted_weights[order]
-        sorted_weights /= sorted_weights.sum()
+            mus = observations
+            if sigmas0 is None:
+                pairs_of_sigma_and_idx = np.empty((n_observations, 2))
+                pairs_of_sigma_and_idx[:, 0] = np.maximum(
+                    pairs_of_observation_and_idx[1:-1, 0] - pairs_of_observation_and_idx[0:-2, 0],
+                    pairs_of_observation_and_idx[2:, 0] - pairs_of_observation_and_idx[1:-1, 0],
+                )
+                pairs_of_sigma_and_idx[:, 1] = pairs_of_observation_and_idx[1:-1, 1]
+                sigmas = pairs_of_sigma_and_idx[np.argsort(pairs_of_sigma_and_idx[:, 1])][:, 0]
+            else:
+                sigmas = sigmas0 * (high - low)
```

Comment:
📝 This `_calculate_numerical_params` is the main diff between the additions in this file and the previous removed multivariate parzen estimator.

Reply:
The logic here seems a bit different even for the univariate case from before. Why was it for instance that we shouldn't sort the observations before inserting the prior? Could you just briefly explain how the algorithms is kept or changes, especially for the univariate case (`sigma0` is `None`)?

Reply:
The algorithm for the univariate case is the same as before; in the multivariate case, the `weight` calculation and the `mus` and `sigmas` calculations are done independently. The implementation is different from before because the order of them must match the order of the `observations` passed to the init of `_ParzenEstimator`. I did the following to keep the algorithm itself unchanged. Pair the observation values with their indices, calculate the sigma using the same algorithm as before looking only at the observations, and then re-sort based on the indices.